include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -0,0 +1,111 @@
+#ifndef _NBL_BUILTIN_HLSL_WORKGROUP_FFT_INCLUDED_
+#define _NBL_BUILTIN_HLSL_WORKGROUP_FFT_INCLUDED_
+
+#include <nbl/builtin/hlsl/subgroup/fft.hlsl>
+#include <nbl/builtin/hlsl/bit.hlsl>
+
+namespace nbl 
+{
+namespace hlsl
+{
+namespace workgroup
+{
+
+template<typename Accessor, typename Scalar, uint32_t WorkgroupSize>
+void store(NBL_REF_ARG(Accessor) sharedmemAccessor, uint32_t threadID, NBL_CONST_REF_ARG(complex_t<Scalar>) lo, NBL_CONST_REF_ARG(complex_t<Scalar>) hi) {
+    sharedmemAccessor.set(threadID + 0 * WorkgroupSize, lo.real());
+    sharedmemAccessor.set(threadID + 1 * WorkgroupSize, hi.real());
+    sharedmemAccessor.set(threadID + 2 * WorkgroupSize, lo.imag());  
+    sharedmemAccessor.set(threadID + 3 * WorkgroupSize, hi.imag());    
+}
+
+template<typename Accessor, typename Scalar, uint32_t WorkgroupSize>
+void load(NBL_CONST_REF_ARG(Accessor) sharedmemAccessor, uint32_t threadID, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi) {
+    lo.real(sharedmemAccessor.get(threadID + 0 * WorkgroupSize));
+    hi.real(sharedmemAccessor.get(threadID + 1 * WorkgroupSize));
+    lo.imag(sharedmemAccessor.get(threadID + 2 * WorkgroupSize));  
+    hi.imag(sharedmemAccessor.get(threadID + 3 * WorkgroupSize));    
+}
+
+template<typename Accessor, typename Scalar, uint32_t WorkgroupSize, bool inverse>
+void FFT_loop(NBL_REF_ARG(Accessor) sharedmemAccessor, uint32_t step, uint32_t threadID){
+    const uint32_t sub_ix = threadID & (step - 1);
+    const uint32_t lo_x_ix = impl::bitfieldInsert(threadID, 0u ,sub_ix, 1u);
+    const uint32_t hi_x_ix = lo_x_ix | step;
+    const uint32_t lo_y_ix = lo_x_ix + 2 * WorkgroupSize;
+    const uint32_t hi_y_ix = hi_x_ix + 2 * WorkgroupSize;
+
+    sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+    complex_t<Scalar> lo = {sharedmemAccessor.get(lo_x_ix), sharedmemAccessor.get(lo_y_ix)};
+    complex_t<Scalar> hi = {sharedmemAccessor.get(hi_x_ix), sharedmemAccessor.get(hi_y_ix)};
+
+    fft::DIX<Scalar, inverse>::radix2(fft::twiddle<Scalar, inverse>(sub_ix, step << 1), lo, hi);
+
+    // Share results between threads
+    sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+    sharedmemAccessor.set(lo_x_ix, lo.real());
+    sharedmemAccessor.set(hi_x_ix, hi.real());
+    sharedmemAccessor.set(lo_y_ix, lo.imag());
+    sharedmemAccessor.set(hi_y_ix, hi.imag());     
```

Comment:
I think there's a way to optimize the shared memory traffic, not only for less load/store, but less smem usage overall.

Lets assume SubgroupSize 32 and Workgroup Size 1024

## When you do a forward FFT of size N=WorkgroupSize*2

Your first step has stride WorkgroupSize and requires no smem transfer.

### Step WorkgroupSize/2

But the second step steps with `WorkgroupSize/2` meaning half of the subgroups load whole workgroup.

Subgroups [0,16) don't actually need to trade their `lo` items, they can stay in registers. You'd be only storing them to load them again. They should only store their `hi` items from previous step.

Subgroups [16,32) don't need to trade their `hi` items, they should only store their `lo` items from previous step.

### Step WorkgroupSize/4

Subgroups [0,8) and [16,24) can stay where they are, and keep their `lo` items, while subgroups [8,16) and [24,32) "move" but can keep their `hi` items.

...

I think you can start to see the pattern emerging.

Bottom "half" of an iteration step bundle stores `hi` to memory, while Top stores `lo`.

Then the Top half gets reassigned.